input_test.go

```diff
@@ -0,0 +1,217 @@
+package underflow
+
+import (
+	"testing"
+
+	"github.com/hexops/autogold"
+	"github.com/onflow/cadence"
+	"github.com/onflow/cadence/runtime/sema"
+	"github.com/stretchr/testify/require"
+)
+
+type InputTestCase struct {
+	want     autogold.Value
+	input    interface{}
+	typeHint sema.Type
+}
+
+var resolver = func(f string, resolverType ResolveType) (string, error) {
+	if f == "first" {
+		return "0x179b6b1cb6755e31", nil
+	}
+	return "", nil
+}
+
+var (
+	stringVal  = "foobar"
+	structType = cadence.StructType{
+		Fields: []cadence.Field{{
+			Identifier: "bar",
+			Type:       cadence.StringType,
+		}},
+	}
+	structTypeAddress = cadence.StructType{
+		Fields: []cadence.Field{{
+			Identifier: "bar",
+			Type:       cadence.AddressType,
+		}},
+	}
+)
+
+func TestParseInputValueWithTypeHint(t *testing.T) {
+	fix64, _ := cadence.NewUFix64("42.1000000")
+	tests := []InputTestCase{
+		{
+			want:     autogold.Want("string", cadence.String("foo")),
+			input:    "foo",
+			typeHint: sema.StringType,
+		},
+		{
+			want:     autogold.Want("uint64", cadence.UInt64(42)),
+			input:    42,
+			typeHint: sema.UInt64Type,
+		},
+		{
+			want:     autogold.Want("uint64 from uint8", cadence.UInt64(42)),
+			input:    uint8(42),
+			typeHint: sema.UInt64Type,
+		},
+
+		{
+			want:     autogold.Want("uint32", cadence.UInt32(42)),
+			input:    42,
+			typeHint: sema.UInt32Type,
+		},
+		{
+			want:     autogold.Want("ufix64", fix64),
+			input:    42.1,
+			typeHint: sema.UFix64Type,
+		},
+		{
+			want:     autogold.Want("optional string", cadence.Optional{Value: cadence.String("foo")}),
+			input:    "foo",
+			typeHint: &sema.OptionalType{Type: sema.StringType},
+		},
+		{
+			want:     autogold.Want("optional string empty", cadence.Optional{}),
+			input:    nil,
+			typeHint: &sema.OptionalType{Type: sema.StringType},
+		},
+		{
+			want:     autogold.Want("address", cadence.Address{23, 155, 107, 28, 182, 117, 94, 49}),
+			input:    "first",
+			typeHint: sema.TheAddressType,
+		},
+		{
+			want:     autogold.Want("stringPointer", cadence.Optional{Value: cadence.String("foobar")}),
+			input:    &stringVal,
+			typeHint: &sema.OptionalType{Type: sema.StringType},
+		},
+		{
+			want:     autogold.Want("string pointer empty", cadence.Optional{Value: cadence.String("foobar")}),
```

Comment:
this test must be wrong, it makes no sense it is basically the same as the one above is it not?

Reply:
```suggestion
			want:     autogold.Want("string pointer empty", cadence.Optional{Value: cadence.String("")}),
```
we want it to be empty string? or should the optional just be completely empty (eg without even the cadence.String value)?

Reply:
maybe` cadence.NewOptional(nil)` . not sure with the syntax

Reply:
i removed this test in my latest commit, but not sure if we should add one with empty string.

Reply:
maybe we can try to keep it if we can, no harm there